Snowflake.Data/Core/RestRequester.cs

```diff
@@ -34,17 +36,87 @@ internal class RestRequester : IRestRequester
     {
         private static SFLogger logger = SFLoggerFactory.GetLogger<RestRequester>();
 
-        private static readonly RestRequester instance = new RestRequester();
+        private static Dictionary<string, IRestRequester> RestRequestersByProxy = 
+            new Dictionary<string, IRestRequester>();
 
-        private RestRequester()
+        private static readonly object requesterPoolLock = new object();
+
+        // The proxy to use making the requests
+        internal IWebProxy Proxy;
+
+        private RestRequester(IWebProxy proxy)
         {
+            Proxy = proxy;
         }
-        
-        static internal RestRequester Instance
+
+        /// <summary>
+        /// Get the RestRequester associated with the given proxy information or create a new one
+        /// if none exist in the pool already.
+        /// </summary>
+        /// <param name="proxyHost">The proxy host.</param>
+        /// <param name="proxyPort">The proxy port.</param>
+        /// <param name="proxyUser">The proxy username or null if none.</param>
+        /// <param name="proxyPassword">The proxy password or null if none.</param>
+        /// <param name="noProxyList">The list of urls to by-pass the proxy.</param>
+        /// <returns>The RestRequester for this proxy.</returns>
+        /// <exception cref="System.FormatException">The port is not a valid int</exception>
+        /// <exception cref="System.OverflowException">The port value is too large</exception>
+        internal static IRestRequester GetRestRequester(
+            string proxyHost,
+            string proxyPort,
+            string proxyUser,
+            string proxyPassword,
+            string noProxyList)
```

Comment:
Can we wrap those string into a single class/struct, maybe name it `ProxyInfo`? It might be easier since you don't need to pass 5 arguments around.

Reply:
I think we should pass in IWebProxy here?

Reply:
That was my original idea too, but it isn't really going to work.
The pool of requesters is global to the driver.
The RestRequester is the one who has the responsibility to create a new IWebProxy based on the connection settings if needed, because it "knowns" about the existing ones.
Where would you create the IWebProxy otherwise? It's an object, to make sure a new one is not created when there is already one for the same settings, you would have to go through the list of existing IWebProxy and check the Address, Credentials and byPassList properties. This is why I use a map of settings/IWebProxy as a fast way to know if there was an existing IWebProxy ready to use.

Reply:
For the first comment, I am not sure that a class/struct is really needed.
It would mean creating a new instance, store the info inside, pass the object instead of the strings themselves and right away extract the strings from the class/struct. 
There is only one call to GetRestRequester in SFSession to get an requester for the Session. This is more or less similar to calling a constructor. The difference here is that it creates a new instance only if needed. I would not create a class/struct for a constructor, would you?
I also like to make it clear, by listing them, that the RestRequester instance returned is based on these 5 settings. If you use a class/struct it hides this a little bit more.